### DISCUSSION TITLE
Latency increase for existing loaded collection after query node scale out

### DISCUSSION THREAD

**Discussion by shawnatmeredith** [milvus-io/milvus#27476] [Q&A and General discussion]
I'm working to reproduce an issue we saw on Milvus 2.3.0 where we saw latency increase after scaling out query node capacity to load additional indexes.  This was a roughly 10x latency increase from 40ms to 400ms on an existing collection that was loaded before scaling.  We had to unload, deindex, and delete the collection then restore from backup to get nominal latency again.

If I can repoduce in an isolated cluster I'll collect logs for sharing here.

**Comment by filip-halt** [comment]
Was this increase permanent? How long were you testing the latencies?

**Reply by shawnatmeredith** [reply]
The increase seemed permanent in the short timeframe before we opted to fully delete and restore the collection.  This was roughly 24 hours where the 10x latency increase persisted.

So far, I have not been able to recreate in an isolated environment.  I'll timebox another few hours early next week to try to reproduce again.

**Comment by xiaofan-luan** [comment]
> I'm working to reproduce an issue we saw on Milvus 2.3.0 where we saw latency increase after scaling out query node capacity to load additional indexes. This was a roughly 10x latency increase from 40ms to 400ms on an existing collection that was loaded before scaling. We had to unload, deindex, and delete the collection then restore from backup to get nominal latency again.
> 
> If I can repoduce in an isolated cluster I'll collect logs for sharing here.

I think this might be due to the growing segment index bug that will be fixed in 2.3.2

**Comment by xiaofan-luan** [comment]
it will also be fixed automatically when index build done.

**Reply by shawnatmeredith** [reply]
Can you elaborate on what you mean by this @xiaofan-luan ?  Is there a link to the growing segment bug you identified?

**Reply by filip-halt** [reply]
Milvus keeps data in segments (normally 512mb). These segments are indexed separately to create many small indexes that are searched in parallel. When inserting data we need to be able to search this data so we keep growing segments which are segments that haven't hit 512mb. Once they hit 512 they are sent to index node to be indexed. Before we used to search these growing segments in a brute force manner, but in 2.3 we added a small index that is grown on the data at the same time. There is a bug (not sure on which one) that might be causing slowdowns/problems with this small index. It goes away when that segment is sent to index node and replaced with a true full index.

**Comment by shawnatmeredith** [comment]
Just pinging back here to confirm that we could not reproduce this so far.  We suspect early data loading scripts, which were performing unnecessary flushes, may have contributed to the decrease rather than the scaling actions taken around that time.

At this point, we are satisfied to move on with our evaluation but will share more info here in the event we run into similar problems in the future.
